source/common/http/http2/codec_impl.cc

```diff
@@ -411,29 +436,38 @@ ssize_t ConnectionImpl::StreamImpl::onDataSourceRead(uint64_t length, uint32_t*
       }
     }
 
-    return std::min(length, pending_send_data_.length());
+    return std::min(length, pending_send_data_->length());
   }
 }
 
-void ConnectionImpl::StreamImpl::onDataSourceSend(const uint8_t* framehd, size_t length) {
+ssize_t ConnectionImpl::StreamImpl::onDataSourceSend(const uint8_t* framehd, size_t length) {
   // In this callback we are writing out a raw DATA frame without copying. nghttp2 assumes that we
   // "just know" that the frame header is 9 bytes.
   // https://nghttp2.org/documentation/types.html#c.nghttp2_send_data_callback
   static const uint64_t FRAME_HEADER_SIZE = 9;
 
+  if (parent_.h2_watermark_improvements_ && length > 0 &&
+      parent_.connection_.aboveHighWatermark()) {
+    // The network connection's output buffer is full. Delay generation of the data frame until the
+    // buffer's size drops to its low-watermark.
+    return NGHTTP2_ERR_WOULDBLOCK;
```

Comment:
Do we have stats counters that track number of frames pending in the current codec implementation?

Reply:
Were we looking for stats, or just to make sure our current frame flood flame logic was unified so network connection frames and nghttp2 frames both count towards the flood limits?

Reply:
I guess my question is if we're losing some of our flooding protection then?
We were using the insta-serialization into data frames to avoid a reflection time OOM attack where someone sent a bunch of requests which would 400 faster than the network drained, queuing outbound header frames.  If I understand this PR correctly those header frames will be buffered by nghttp2 and we won't account for them.  Doesn't that reopen that attack vector?

Reply:
We don't enforce a byte limit on response headers.  What we enforce is a count on number of outstanding control frames.  The limit on outstanding control frames is still enforced correctly; we sum the number of frames that are serialized and the number of frames internally queued by nghttp2.

So I think we're not losing any flood protections.  Http2BufferWatermarksTest.PingFloodAfterHighWatermark provides test coverage.  We could add a similar test with header frames.

Reply:
I do no think we loose any flood protection. The frames that are accumulated by nghttp2 are accounted for using the `Nghttp2Session::getOutboundControlFrameQueueSize()` method when making flood checks.
I was concerned about the visibility into the size of the queue maintained by nghttp2. We do have counters for buffer sizes (not number of frames) and now we will have some number of bytes sitting in the nghttp2 queue without them being reflected in any stats.